include/alpaka/mem/buf/BufOacc.hpp

```diff
@@ -0,0 +1,652 @@
+                    static_assert(
+                        std::is_same<TIdx, Idx<TExtent>>::value,
+                        "The idx type of TExtent and the TIdx template parameter have to be identical!");
+                }
+
+            public:
+                DevOacc m_dev;
+                Vec<TDim, TIdx> m_extentElements;
+                Vec<TDim, TIdx> m_pitchBytes;
+                TElem* m_pMem;
+
+                    BufOaccImpl(const BufOaccImpl&) = delete;
+                    BufOaccImpl(BufOaccImpl &&) = default;
+                BufOaccImpl& operator=(const BufOaccImpl&) = delete;
+                BufOaccImpl& operator=(BufOaccImpl&&) = default;
+                    ~BufOaccImpl()
+                {
+                    m_dev.makeCurrent();
+                    acc_free(m_pMem);
+                }
+            };
+        }
+    }
+
+    template<
+        typename TElem,
+        typename TDim,
+        typename TIdx>
+    class BufOacc
+    {
+    public:
+        //-----------------------------------------------------------------------------
+        //! Constructor
+        template<
+            typename TExtent>
+        ALPAKA_FN_HOST BufOacc(
+            DevOacc const & dev,
+            TElem * const pMem,
+            TExtent const & extent) :
+                m_spBufImpl(std::make_shared<oacc::detail::BufOaccImpl<TElem, TDim, TIdx>>(dev, pMem, extent))
+        {}
+
+            BufOacc(const BufOacc&) = default;
+            BufOacc(BufOacc &&) = default;
+        BufOacc& operator=(const BufOacc&) = default;
+        BufOacc& operator=(BufOacc&&) = default;
+
+        oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() {return *m_spBufImpl;}
+        const oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() const {return *m_spBufImpl;}
+
+        inline const Vec<TDim, TIdx>& extentElements() const {return m_spBufImpl->m_extentElements;}
+
+    private:
+        std::shared_ptr<oacc::detail::BufOaccImpl<TElem, TDim, TIdx>> m_spBufImpl;
+    };
+
+    namespace traits
+    {
+        //#############################################################################
+        //! The BufOacc device type trait specialization.
+        template<
+            typename TElem,
+            typename TDim,
+            typename TIdx>
+        struct DevType<
+            BufOacc<TElem, TDim, TIdx>>
+        {
+            using type = DevOacc;
+        };
+        //#############################################################################
+        //! The BufOacc device get trait specialization.
+        template<
+            typename TElem,
+            typename TDim,
+            typename TIdx>
+        struct GetDev<
+            BufOacc<TElem, TDim, TIdx>>
+        {
+            ALPAKA_FN_HOST static auto getDev(
+                BufOacc<TElem, TDim, TIdx> const & buf)
+            -> DevOacc
+            {
+                return (*buf).m_dev;
```

Comment:
That looks very strange to me, since `buf` is a reference. So `BufOacc<TElem, TDim, TIdx>` models some kind of pointer type?

Reply:
@jkelling